### PR TITLE
Update sparkle to work with latest jni.

### LINKED ISSUE #85: Use foreignptr values to cleanup global references in finalizers

### PR DESCRIPTION
**Pull request by facundominguez**
jni now use ForeignPtr's for java references.

Closes #85

### CODE REVIEW

File: src/Control/Distributed/Spark/Closure.hs [hunk 1]
```diff
@@ -22,27 +22,35 @@ module Control.Distributed.Spark.Closure
...
 -- function is foreign exported.
+--
+-- The function in the closure pointed by the first argument must yield
+-- a local reference to a Java object, or the reference might be released
+-- prematurely.
 apply
-  :: JByteArray
-  -> JObjectArray
-  -> IO JObject
+  :: Ptr JByteArray
```
Comment (mboes) [inline src/Control/Distributed/Spark/Closure.hs]: This change in type signature makes it look like you're adding an extra level of indirection? But are you? It looks like you can simply unwrap `J`.

Comment (facundominguez) [inline src/Control/Distributed/Spark/Closure.hs]: This type is what `J <$> newForeignPtr_ bytes` expects below. No indirection implied.

Comment (mboes) [inline src/Control/Distributed/Spark/Closure.hs]: Ok so let's keep the original signature.

**Comment by facundominguez** [comment]
It doesn't look like I can honor the request to keep the old signature. FFI bindings can't have arguments of type `ForeignPtr`.

**Comment by mboes** [comment]
Ok fair enough.

**Pull request merged**